### ISSUE TITLE
virtiofsd crashes on Ubuntu 16.04 due to failing seccomp_init

### ISSUE THREAD

**Issue by gkunz** [kata-containers/runtime#2829]
# Description of problem

I have been trying to run a simple test container with Kata and CLH on Ubuntu 16.04.6 LTS. Kata has been deployed using the kata-deploy daemonset.

Problem: The pod remains in ContainerCreating state indefinitely, due to a crashing virtiofsd.

The test pod definition looks like this

```
apiVersion: v1
kind: Pod
metadata:
  name: ubuntu-untrusted
spec:
  runtimeClassName: kata-clh
  containers:
    - name: ubuntu
      image: ubuntu
```

virtiofsd crashes on Ubuntu 16.04.6 LTS due to failing seccomp_init syscalls.

# Expected result

The expected result is that the pod comes up.

# Actual result

From syslog:

```
Jul 14 14:11:38 pod2-blade08 kata[17620]: time="2020-07-14T14:11:38.483440052+02:00" level=debug msg="virtiofsd: seccomp_init(): Operation not supported" ID=46364129b31b158a32dc750862567756fff5077e3f437e7dcde767c242068c7d source=virtiofsd
Jul 14 14:11:38 pod2-blade08 containerd[14567]: time="2020-07-14T14:11:38.483440052+02:00" level=debug msg="virtiofsd: seccomp_init(): Operation not supported" ID=46364129b31b158a32dc750862567756fff5077e3f437e7dcde767c242068c7d source=virtiofsd
```

In fact, after every crash of virtiofsd, a new instance of clh is spawned, resulting in a quickly increasing number of cloud-hypervisor instances:

```
kata (master #)]$ ps -eaf |grep cloud
root     12365     1  0 22:00 ?        00:00:00 /opt/kata/bin/cloud-hypervisor --api-socket /run/vc/vm/420dbcc002f74d992ba541c5f9a6e2ab5bf4cddf20d83c644bab10d5c92014aa/clh-api.sock --log-file /run/vc/vm/420dbcc002f74d992ba541c5f9a6e2ab5bf4cddf20d83c644bab10d5c92014aa/clh.log
root     12647     1  0 22:00 ?        00:00:00 /opt/kata/bin/cloud-hypervisor --api-socket /run/vc/vm/60b2d1423b31a7a38ab24f073a8986af02f81592eda58f76169418ca0099bae7/clh-api.sock --log-file /run/vc/vm/60b2d1423b31a7a38ab24f073a8986af02f81592eda58f76169418ca0099bae7/clh.log
root     12849     1  0 22:00 ?        00:00:00 /opt/kata/bin/cloud-hypervisor --api-socket /run/vc/vm/a2bd8ac0f878de90c60a14718e75206aa39186616a5e44b75a6bf016aa9903ea/clh-api.sock --log-file /run/vc/vm/a2bd8ac0f878de90c60a14718e75206aa39186616a5e44b75a6bf016aa9903ea/clh.log
root     13083     1  0 22:00 ?        00:00:00 /opt/kata/bin/cloud-hypervisor --api-socket /run/vc/vm/d8fc12cbac59e8e6bb64d9f5b8cde75d659b2005d7f4a49e91dc68d77b8c2fff/clh-api.sock --log-file /run/vc/vm/d8fc12cbac59e8e6bb64d9f5b8cde75d659b2005d7f4a49e91dc68d77b8c2fff/clh.log
root     13292     1  0 22:00 ?        00:00:00 /opt/kata/bin/cloud-hypervisor --api-socket /run/vc/vm/7dc9eb8a2f1a6583032a1cae3959f9c94c79556b0a61c392797205e7f97aca5e/clh-api.sock --log-file /run/vc/vm/7dc9eb8a2f1a6583032a1cae3959f9c94c79556b0a61c392797205e7f97aca5e/clh.log
root     13563     1  0 22:00 ?        00:00:00 /opt/kata/bin/cloud-hypervisor --api-socket /run/vc/vm/62a001833a7fdc51f5e8cf3aa2097f849f5420a075ab639e46ec0a6023e3e039/clh-api.sock --log-file /run/vc/vm/62a001833a7fdc51f5e8cf3aa2097f849f5420a075ab639e46ec0a6023e3e039/clh.log
[...]
```

```
kata (master #)]$ ps -eaf | grep cloud | wc -l
111
```

I can successfully resolve the problem by building virtiofsd manually on my system.

**Comment by c3d** [comment]
The way I see it, we have to somewhat distinct issues here:
* virtiofsd crashing, which you tracked down to missing syscall
* the respawn behavior and lack of error reporting.

Would you mind opening another issue for the second one?

**Comment by dagrh** [comment]
My bet is it's missing SCMP_ACT_KILL_PROCESS in the kernel but not the build env.
Our current setup_seccomp has:
#ifdef SCMP_ACT_KILL_PROCESS
    ctx = seccomp_init(SCMP_ACT_KILL_PROCESS);
    /* Handle a newer libseccomp but an older kernel */
    if (!ctx && errno == EOPNOTSUPP) {
        ctx = seccomp_init(SCMP_ACT_TRAP);
    }
#else
    ctx = seccomp_init(SCMP_ACT_TRAP);
#endif
    if (!ctx) { 
        fuse_log(FUSE_LOG_ERR, "seccomp_init() failed\n"); 
        exit(1);
    }

I think that fall back to SCMP_ACT_TRAP went in because of this issue.

**Comment by gkunz** [comment]
@c3d yes, you are right. I'll create a separate issue for the respawn behavior.

**Comment by dagrh** [comment]
@gkunz It would be great if you could try https://gitlab.com/virtio-fs/qemu/-/tree/qemu5.0-virtiofs-dax  to see if it fixed it for you.

**Comment by gkunz** [comment]
@dagrh: ok, I'll give that a try and report back.

As I was intending to run workloads on cloud-hypervisor, I came across the cloud-hypervisor documentation regarding virtiofsd [1]. I eventually used the code of the "private" repo listed there to build virtiofsd locally and this worked - after making sure that the CLI options Kata uses for virtiofsd are compatible with this version of virtiofsd... Is the repo you linked above the "official repo" for virtiofsd?

[1] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/docs/fs.md

**Comment by gkunz** [comment]
@dagrh One issue I immediately ran into when using the repo and branch you mentioned is a mismatch between the command line options expected by virtiofsd and the command line options used by the stable release of Kata I am getting with the kata-deploy daemonset (v1.11.1): the kata runtime still uses the "-o vhost_user_socket" option while virtiofsd has dropped this option [1]. However, this seems to be fixed in 1.12.0-alpha.

Moreover, when switching to Kata 1.12.0-alpha, the version of virtiofsd of [3] works with qemu and CLH, but with the version you pointed to, the pods get stuck in ContainerCreating. I haven't digged deeper into this yet.

In summary, it has been a little confusing/challenging o find the right combination of virtiofsd and kata to make stuff work - at least on Ubuntu 16.04. As I mentioned, I personally don't have a strong preference on 16.04.

[1] https://github.com/kata-containers/runtime/blob/1.11.2/virtcontainers/virtiofsd.go#L132
[2] https://github.com/kata-containers/runtime/blob/1.12.0-alpha0/virtcontainers/virtiofsd.go#L105
[3] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/docs/fs.md

**Comment by ariel-adam** [comment]
@dagrh Any comments on this one?

**Comment by dagrh** [comment]
The current Kata and ucrrent QEMU work together OK these days; they have matching command lines.
The actual original 16.04 related crash was fixed ages ago (falling back to SCMP_ACT_TRAP even if the headers have ACT_KILL_PROCESS).
Mixing a virtiofsd with a cloudhypervisor is probably a more interesting question; I'd stick with the rust replacement for virtiofsd when using CH.

**Comment by fidencio** [comment]
---

This issue is being automatically closed as Kata Containers 1.x has now reached EOL (End of Life). This means it is **no longer being maintained**.

> **Important:**
>
> All users should switch to the latest [Kata Containers 2.x release](https://github.com/kata-containers/kata-containers/blob/main/docs/install/README.md) to ensure they are using a maintained release that contains the latest security fixes, performance improvements and new features.

This decision was discussed by the @kata-containers/architecture-committee and has been announced via the [Kata Containers mailing list](http://lists.katacontainers.io/pipermail/kata-dev/):
* http://lists.katacontainers.io/pipermail/kata-dev/2020-November/001601.html
* http://lists.katacontainers.io/pipermail/kata-dev/2021-April/001843.html
* http://lists.katacontainers.io/pipermail/kata-dev/2021-May/001896.html

If you believe this issue still applies to Kata Containers 2.x, please open an issue against the [Kata Containers 2.x repository](https://github.com/kata-containers/kata-containers), pointing to this one, providing details to allow us to migrate it.

---
